### ISSUE TITLE
Proposed ENH: Allow user to turn off breaking of streamlines in streamplot

### ISSUE THREAD

**Issue by ketch** [matplotlib/matplotlib#8388]
**tl;dr**: streamplot does this:
![download](https://cloud.githubusercontent.com/assets/697763/24358697/d3bd7a88-12b6-11e7-97b5-fbb935c17074.png)

I'd like this:
![download1](https://cloud.githubusercontent.com/assets/697763/24358639/a48f39cc-12b6-11e7-9879-1f8e1205d945.png)

I found a way to do this and I would like to submit a PR.  I'm asking for input from a developer on whether this would be worthwhile and whether the approach I'm thinking of makes sense.

## More details

I'm plotting a vector field with streamplot, and no matter how large I set the density, it interrupts streamlines at (seemingly random) points.  My understanding is that streamplot uses a grid and only allows one streamline to pass through each cell of the grid.  So if streamlines get too close to each other, one of them must end.

What I'd really like is that if any part of a given streamline is plotted, then the whole thing is plotted.  That is, streamlines should continue in both directions until they leave the plot region.  You might think that you could achieve this by setting a large value of the `density` argument.  But it seems that `density` *also* determines how many starting points are selected for streamlines, so this doesn't help.  I'm using Matplotlib 2.0.0 but I've also tried with a clone of the current dev version.

I've found a way to accomplish what I want, by simply changing the last line of this snippet (from the StreamMask class):

```python
    def _update_trajectory(self, xm, ym):
        """Update current trajectory position in mask.

        If the new position has already been filled, raise `InvalidIndexError`.
        """
        if self._current_xy != (xm, ym):
            if self[ym, xm] == 0:
                self._traj.append((ym, xm))
                self._mask[ym, xm] = 1
                self._current_xy = (xm, ym)
            else:
                raise InvalidIndexError
```
to
```python
    def _update_trajectory(self, xm, ym):
        """Update current trajectory position in mask.

        If the new position has already been filled, raise `InvalidIndexError`.
        """
        if self._current_xy != (xm, ym):
            if self[ym, xm] == 0:
                self._traj.append((ym, xm))
                self._mask[ym, xm] = 1
                self._current_xy = (xm, ym)
            else:
                pass
```

I'd like to submit a PR that allows the user to do this.  My suggestion would be:

- Have a keyword argument `no_broken_streamlines` (or similar) that defaults to False.
- If set to True, then streamplot will behave as with my patch above.

To reproduce the figure at the top, use this code:

```python
import numpy as np
import matplotlib.pyplot as plt

N = 400
h, hu = np.meshgrid(np.linspace(0.1,6,N),np.linspace(-3,3,N))
g = 1.
u = hu/h
dh = np.ones_like(h)
dhu = u - np.sqrt(g*h)
plt.streamplot(h,hu,dh,dhu,density=1.,arrowstyle='-')
plt.axis('tight');
```

**Comment by lzhadid** [comment]
Hi,
is this the file you modified ? 
https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/streamplot.py

What did you put for "axes" when using the streamplot function ?

**Comment by ketch** [comment]
> is this the file you modified ?
> https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/streamplot.py

Yes.

>What did you put for "axes" when using the streamplot function ?

I don't think I passed an `axes` argument.

**Comment by lzhadid** [comment]
I tried without passing the axes argument, it didn't work. I had to define the "axes", it is actually an input in the defined function:

def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
               cmap=None, norm=None, arrowsize=1, arrowstyle='-|>',
               minlength=0.1, transform=None, zorder=None, start_points=None,
               maxlength=4.0, integration_direction='both'):
    """Draws streamlines of a vector flow.
    *x*, *y* : 1d arrays
        an *evenly spaced* grid.

**Comment by afvincent** [comment]
@lzhadid *axes* is the `Axes` instance where you want to plot the streamplot. You can use `plt.gca()` to get the lastly used `Axes` instance when you are using the pyplot interactive interface.

**Comment by lzhadid** [comment]
Thank you! It works now. It wasn't working because I was using fig0, ax = plt.subplots(). Using only ax=plt.gca() works perfectly!

**Comment by lzhadid** [comment]
Do you know if one can get the coordinates of the different streamlines in an ordered way, other than .lines.get_segments() or .lines.get_paths() ?

**Comment by tacaswell** [comment]
@ketch Sorry this fell through the cracks in March!

Can you open a Pull Request with those changes + a way to make this behavior configurable through the `streamplot` api?

**Comment by Cyclone-Zhang** [comment]
It's not working now by changing the last line of < _update_trajectory>

**Comment by henry2004y** [comment]
Is there any update to this issue? Years have passed and I am still suffering from the same thing...

**Comment by timhoffm** [comment]
Sorry, no update on this. As you can see from issues and open PRs, matplotlib is a large project with limited resources. Any updates would have been noted here.

**Comment by ketch** [comment]
@henry2004y I opened a PR -- if you're interested, please complete the outstanding parts there.  Or you can just check out my branch and it will work (just pass broken_streamlines=False when calling streamplot).

**Comment by henry2004y** [comment]
Thanks for sharing! Actually I noticed that in Matlab the default streamline has the same behavior as the original Matplotlib version. I looked further into some other libraries like yt and SpacePy, and they seem to have this implemented, but just for their internal data structure. Even larger visualization software like Tecplot and ParaView for CFD have streamline implemented as we want, but sometimes they are not so easy to use through APIs.

**Comment by donglai96** [comment]
> @henry2004y I opened a PR -- if you're interested, please complete the outstanding parts there. Or you can just check out my branch and it will work (just pass broken_streamlines=False when calling streamplot).

Thank you, this saved my day. I will find time to work on the remaining PR issue

**Comment by oscargus** [comment]
Fixed by https://github.com/matplotlib/matplotlib/pull/22707
